File: samples/JavaSamples/src/main/java/com/mku/salmon/samples/samples/DataStreamSample.java
package com.mku.salmon.samples.samples;

import com.mku.convert.BitConverter;
import com.mku.salmon.streams.EncryptionMode;
import com.mku.salmon.streams.SalmonStream;
import com.mku.streams.MemoryStream;

import java.io.IOException;
import java.util.Arrays;

public class DataStreamSample {
    static int BUFFER_SIZE = 256 * 1024; // recommended buffer size aligned to internal buffers

    public static byte[] encryptDataStream(byte[] data, byte[] key, byte[] nonce) throws IOException {
        System.out.println("Encrypting bytes: " + BitConverter.toHex(Arrays.copyOf(data, 24)) + "...");

        // we use a memory stream to host the encrypted data
        byte[] encData = new byte[data.length];
        MemoryStream memoryStream = new MemoryStream(encData);

        // and wrap it with a SalmonStream that will do the encryption
        SalmonStream encStream = new SalmonStream(key, nonce, EncryptionMode.Encrypt, memoryStream);

        // now write the data you want to decrypt
        // it is recommended to use a large enough buffer while writing the data
        // for better performance
        int totalBytesWritten = 0;
        while (totalBytesWritten < data.length) {
            int length = Math.min(data.length - totalBytesWritten, DataStreamSample.BUFFER_SIZE);
            encStream.write(data, totalBytesWritten, length);
            totalBytesWritten += length;
        }
        encStream.flush();

        // the encrypted data are now written to the memoryStream/encData.
        encStream.close();
        memoryStream.close();

        System.out.println("Bytes encrypted: " + BitConverter.toHex(Arrays.copyOf(encData, 24)) + "...");
        return encData;
    }

    public static byte[] decryptDataStream(byte[] data, byte[] key, byte[] nonce) throws IOException {
        System.out.println("Decrypting bytes: " + BitConverter.toHex(Arrays.copyOf(data, 24)) + "...");

        // we use a stream that contains the encrypted data
        MemoryStream memoryStream = new MemoryStream(data);

        // and wrap it with a salmon stream to do the decryption
        SalmonStream decStream = new SalmonStream(key, nonce, EncryptionMode.Decrypt, memoryStream);

        // decrypt the data
        byte[] decData = new byte[(int) decStream.length()];
        int totalBytesRead = 0;
        int bytesRead = 0;
        while ((bytesRead = decStream.read(decData, totalBytesRead, DataStreamSample.BUFFER_SIZE)) > 0) {
            totalBytesRead += bytesRead;
        }

        decStream.close();
        memoryStream.close();

        System.out.println("Bytes decrypted: " + BitConverter.toHex(Arrays.copyOf(decData, 24)) + "...");
        return decData;
    }
}

File: samples/PythonSamples/samples/data_stream_sample.py
#!/usr/bin/env python3
from simple_io.convert.bit_converter import BitConverter
from salmon_core.salmon.streams.encryption_mode import EncryptionMode
from salmon_core.salmon.streams.aes_stream import AesStream
from simple_io.streams.memory_stream import MemoryStream
from simple_io.streams.random_access_stream import RandomAccessStream


class DataStreamSample:
    BUFFER_SIZE = 256 * 1024  # recommended buffer size aligned to internal buffers

    @staticmethod
    def encrypt_data_stream(data, key, nonce):
        print("Encrypting bytes: " + BitConverter.to_hex(data[0:24]) + "...")

        # we use a memory stream to host the encrypted data
        memory_stream: MemoryStream = MemoryStream()

        # and wrap it with a AesStream that will do the encryption
        enc_stream = AesStream(key, nonce, EncryptionMode.Encrypt, memory_stream)

        # now write the data you want to decrypt
        # it is recommended to use a large enough buffer while writing the data
        # for better performance
        total_bytes_written = 0
        while total_bytes_written < len(data):
            length = min(len(data) - total_bytes_written, DataStreamSample.BUFFER_SIZE)
            enc_stream.write(data, total_bytes_written, length)
            total_bytes_written += length
        enc_stream.flush()

        # the encrypted data are now written to the memory_stream/enc_data.
        enc_stream.close()
        enc_data: bytearray = memory_stream.to_array()
        memory_stream.close()

        print("Bytes encrypted: " + BitConverter.to_hex(enc_data[0:24]) + "...")
        return enc_data

    @staticmethod
    def decrypt_data_stream(data, key, nonce):
        print("Decrypting bytes: " + BitConverter.to_hex(data[0:24]) + "...")

        # we use a stream that contains the encrypted data
        memory_stream: MemoryStream = MemoryStream(data)

        # and wrap it with a salmon stream to do the decryption
        dec_stream: RandomAccessStream = AesStream(key, nonce, EncryptionMode.Decrypt, memory_stream)

        # decrypt the data
        dec_data = bytearray(dec_stream.get_length())
        total_bytes_read = 0
        while (bytes_read := dec_stream.read(dec_data, total_bytes_read, DataStreamSample.BUFFER_SIZE)) > 0:
            total_bytes_read += bytes_read

        dec_stream.close()
        memory_stream.close()

        print("Bytes decrypted: " + BitConverter.to_hex(dec_data[0:24]) + "...")
        return dec_data
